rfcs/atomic_record_fields.md

```diff
@@ -0,0 +1,240 @@
+copying the value into a new atomic, which has a completely different
+(and useless and inefficient) semantics.
+
+## The proposal: phantom-typed first-class atomic fields
+
+In addition to the `atomic` keyword (or builtin attribute) on record
+fields, we propose to add:
+
+1. a new type `('r, 'a) Atomic.Field.t`, which denotes a field of type
+  `'a` within a record of type `'r`. Its representation is just an
+  integer, the offset of the field within the record structure.
+  
+2. Atomic operations taking both a `'r` and a `('r, 'a) t` in this new
+   Atomic.Field module, which mimick/duplicate the operations on
+   `Atomic.t`.
+  
+3. A single builtin-in construct to construct `('r, 'a) Atomic.Field.t`
+   values out of atomic record field names: `[%atomic.field foo]`
+
+For example, if we have `type 'a t = { atomic mutable state : 'a st }`,
+then `[%atomic.field state]` is a polymorphic constant of type
+`('a t, 'a st) Atomic.Field.t`.
+
+The `Atomic.Field` submodule looks like this:
+
+```ocaml
+module Field = sig
+  type ('r, 'a) t (* internally: a field offset *)
+  val get : 'r -> ('r, 'a) t -> 'a
+  val compare_and_set : 'r -> ('r, 'a) t -> 'a -> 'a -> bool
+  val fetch_and_add : 'r -> ('r, int) t -> int -> int
+  ... (* other atomic operations *)
+end
+```
+
+The library functions would be implemented as primitives. They can be
+implemented efficiently with this interface.
+
+
+## Advanced topics
+
+### Type disambiguation
+
+One remaining design question is: if `[%atomic.field foo]` refers to
+the field name `foo` at a type in the context, how can we disambiguate
+if there are different record types with such a field in scope? One
+can use module prefixing `[%atomic.field M.foo]`, but even this may be
+ambiguous when a single module contains several record declarations.
+
+Broadly our intuition is that this should be checkable like a field
+access, using propagated type information. In particular, it should
+always be possible to write explicitly:
+`([%atomic.field M.foo] : (my_record, _) Atomic.Field.t`.
+
+It may turn out necessary in practice to support a more concise syntax
+such as `[%atomic.field (M.foo : my_record -> _)]` where the type
+annotation must be of the form `<record type> -> <element type>`. But
+we would suggest trying first without this extra notation.
+
+### Bonus feature: atomic arrays
+
+We can expose atomic arrays with the same mechanism: a builtin type
+`'a Atomic.Array.t`, with a primitive function to build an index from
+an integer:
+
+```ocaml
+module Atomic : sig
+  ....
+  type 'a array
+  val index : int -> ('a t, 'a) Field.t
+end
+```
+
+We can reuse the `Field.t` type and its atomic-operation API, but note
+that the index here may be outside the bounds of the array, and will
+have to be checked on access. (The index is not bound-checked by the
+`index` function, as it may be called on different arrays later.) For
+`Field.t` values at record type, no such bound checking is necessary.
```

Comment:
I’m reminded (by @polytypic) that non-bound-checked accesses would be desirable as well for performance-critical code.

Reply:
Thinking more about this, I notice that there is a weakness in my proposal related to arrays. When using `Atomic.Field.t` for records, we do not want any bound-checking, as the indices are in-bounds by construction. So the `Atomic.Field.t` primitives will not perform any bound-checking. We can reuse the same primitives to implement array access, but the specific proposal I made in the PR leads to non-bound-checked array accesses by default, not to bound-checked array accesses. I think this is wrong: I'm happy to offer both, but bound-checked accesses should be the default.

I see two approaches to solve this:

1. We could of course expose different primitives in `Atomic.Array.t`, which would perform bound-checking first and then call the same runtime primitives as those of the `Atomic.Field.t` API. (More precisely: for each access function, offer a checked and an unchecked version in `Atomic.Array.t`.) I think that this would be perfectly reasonable, and/but it duplicates the API surface which I thought we could avoid.
2. We could change our `index` function to have `index : 'a t -> int -> ('a Array.t, 'a) Field.t`, and also `val unsafe_index : int -> ('a Array.t, 'a) Field.t`, wich `index` in charge of doing bound-checking against a specific array. This is a bit wonky: the index you get has been checked against one array and it is in fact unsafe/unchecked to use it against another array. So only the fragment `Atomic.Field.<op> arr (Atomic.Array.index arr i)` is safe, where you use the index for indexing right away, on the same array. This avoids the API duplication but the API is a bit weird. This would be acceptable for an advanced, expert-users-seeking-performance-only internal API, but I am not fond of the extra complexity (it is more work to explain and to use), so I would rather support (1).

My gut feeling is thus that the best move would be to remove this paragraph on supporting arrays, which was tentative and does not pan out as I had hoped, and simply bite the bullet and offer a `Atomic.Array` submodule without any of the field stuff getting involved. (... and we could also have `Atomic.ContendedArray` if we wanted to, for a fourth copy of the list of primitives.)

Reply:
> I’m reminded (by @polytypic) that non-bound-checked accesses would be desirable as well for performance-critical code.

See  [Saturn's spsc_queue](https://github.com/ocaml-multicore/saturn/blob/main/src_lockfree/spsc_queue.ml) for  example. All accesses are made with an index modulo the size of the array, so there is no need for a bound check.